### DISCUSSION TITLE
Editing the text in PDF

### DISCUSSION THREAD

**Discussion by priyamharsh14** [pymupdf/PyMuPDF#1019] [Looking for help]
I was looking for a way to edit or replace some text inside a PDF file. I read some suggestions on #257, but I am unable to perform it. When I extracted xref stream, there were some `Tj` and `TJ`, and the text data was inside `<>` brackets, but when I tried to decode it (hex to string), the decoded text is not valid.

Example,
`<000e001c00140003001b0019001a0017001b001b0017001300130014> Tj`

I am unable to find the text data stream which I want to replace.

**Comment by JorjMcKie** [comment]
It **_cannot_** work this way!
The hex chars you see there are (in the vast majority of cases) the result of a quite complex translation process, which involves a table (called "CMAP") of the used font.
So you would have to employ redaction annotations for this. The basic approach being:
1. Locate the text, say "needle".
2. Add a redaction annotation with the hit bbox as parameter and apply annotations. This will remove "needle".
3. Insert "newtext" at the location of "needle".

There are a number of additional complexities coming in:
* Use of the same font? If yes, consider that it may be a subset font, which may not have all characters contained in "newtext".
* How long is "newtext"? If shorter or longer than "needle", the result will look ugly.
* "needle" could have been (partly) invisible or transparent. This information cannot be extracted right now. The replacement "newtext" will be definitely visible! The text color however **_can_** be extracted and reused for "newtext".
* ...

**Reply by priyamharsh14** [reply]
Okay. Thank you for your detailed answer.

**Reply by JorjMcKie** [reply]
then you did not correctly locate the text, or you did not invoke `page.apply_redactions()`

**Reply by priyamharsh14** [reply]
Yes, I forgot to apply the redactions. My bad.
